### PR TITLE
feat: Add extended explain info to Comet plan

### LINKED ISSUE #253: Support extended explain plan information
### What is the problem the feature request solves?

https://issues.apache.org/jira/browse/SPARK-47289 adds a new interface to allow extensions to provide additional plan information in spark's explain plan. This information also becomes available in the Spark UI's  SQL tab
An extension can add any information it wants to.
For Comet, it is very useful to track why a particular plan did not enable a Comet operator or which expressions are commonly used by not supported.

### Describe the potential solution

To enable this, during planning, we need to track the reason why a particular operation was not enabled for Comet, bubbling the information all the way up to the root Spark plan.
One way to include this in the Spark plan is to add a Comet specific tag in the Spark plan and  attach this information to the tag. If a plan is 'copied' as is done many times during planning, spark copies any tags over to the transformed plan and the information is then preserved.

### Additional context

The spark trait to enable this is available only from Spark 4.0 so the implementation must work even if the feature is not available in Spark.

### PR DESCRIPTION
**Pull request by parthchandra**
## Which issue does this PR close?

Closes #253

## Rationale for this change

Adds addition planning information to a Spark plan which allows us to track the reasons why a Spark plan was not fully converted to Comet.

## What changes are included in this PR?

This PR adds a `CometExplainInfo` structure that is returned by every step of the planning process. Eventually the Structure is attached to the Spark plan being returned by adding it as a Comet specific tag

## How are these changes tested?

Additional unit test and tested with all TPCH and TPCDS queries with both Spark 3.4 and Spark 3.4 with the extended plan support.
Also regenerated the `CometTPC*QueriesList` output for both TPCH and TPCDS run against a 1TB dataset.
Note: Since the extended plan support does not exist before Spark 4, the output of `CometTPC*QueriesList` was updated with a build of Spark 3.4 with the feature backported.

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/CometExplainInfo.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometExplainInfo.scala:20]: Unrelated to this PR, but we should probably be using `org.apache.datafusion.comet` once the repositories get renamed?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:108]: We could use Scala's `Option` type here rather than defining our own `none` constant. For example:

```scala
            val info1 = if (isSchemaSupported(requiredSchema)) {
              Some(CometExplainInfo(s"Schema $requiredSchema is not supported"))
            } else {
              None
            }
```

We would also need to update the call to pass the list of reasons to `opWithInfo` to use `flatten` so that we only pass the valid reasons.

```scala
opWithInfo(scanExec, CometExplainInfo("SCAN", Seq(info1, info2, info3).flatten))
```

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:490]: Do we need to file an issue to track this?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:699]: very minor nit: we could combine these if we want to and avoid the extra variable `b`.

```suggestion
            case _: CometExec | _: CometBroadcastExchangeExec | _: CometShuffleExchangeExec => op
```

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (dbtsai) [inline spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:56]: For longer lambda expression, we can ignore parentheses. Why do we have a `map(t => t)`?
```    
sorted.foreach { p => 
      val s = getActualPlan(p).getTagValue(CometExplainInfo.EXTENSION_INFO).map(t => t).getOrElse("")
      if (s.nonEmpty) {
        info = info :+ s
      }
    }
```

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (dbtsai) [inline spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:73]: ditto, remove the unneeded parentheses.

File: spark/src/main/scala/org/apache/comet/CometExplainInfo.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometExplainInfo.scala:20]: Agree. But this won't be the only class needing that change. So perhaps better to wait until then and do one bulk update?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:108]: Good idea. I may have a cleaner way of doing this thoguh, which would make this unnecessary. Give me a little time since I need to change a lot of places to try out the idea. If that doesn't work, I'll change this to use Scala `Option`

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala]: This got copied from some other branch so I've removed it. I'm not too sure what else we need to support for `CometTakeOrderedAndProjectExec` but let me find out and log an issue

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala]: changed

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala]: chnaged

File: spark/src/main/scala/org/apache/comet/CometExplainInfo.scala
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometExplainInfo.scala:20]: Yes, absolutely. The move will likely happen next week, so let's discuss after that.

**Comment by parthchandra** [comment]
@andygrove I changed the core of the implementation. Instead of setting information in a CometExplainInfo structure and bubbling it up in the plan, I now set the explain information as a spark tag in the plan or expression.  This makes the changes easier to implement as we add support for more operators and expressions. 
Please take a look.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:107]: I wonder if we could add a helper method to reduce the boilerplate for each of these sections for creating the `Option[String]`.

We could add a method like this:

```scala
def createMessage(condition: Boolean, message: () => String): Option[String] = {
  if (condition) {
    Some(message())
  } else {
    None
  }
}
```

The call site could then be a one-liner:

```scala
val info1 = createMessage(isSchemaSupported(requiredSchema), s"Schema $requiredSchema is not supported")
```

What do you think?

Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:107]: edit: I just updated the suggested code to use the pass-by-name `message: => String` syntax so that the message is evaluated lazily

File: spark/inspections/CometTPCHQueriesList-results.txt
Comment (parthchandra) [reply spark/inspections/CometTPCHQueriesList-results.txt]: Great catch! I was catching this but in FilterExec I was not propagating the correct error. Fixed.

File: spark/inspections/CometTPCHQueriesList-results.txt
Comment (parthchandra) [reply spark/inspections/CometTPCHQueriesList-results.txt]: Another good catch! I reviewed everything and found a few more paths where the error was not being caught. Fixed.

File: spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala
Comment (parthchandra) [reply spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala]: Good point. Done

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala]: Done.

**Comment by parthchandra** [comment]
@advancedxy, @andygrove Addressed your comments. I was missing a few paths where the explanation was not being propagated correctly. I've fixed the ones I could find.
Also, updated the results in CometTPC*QueriesList. I had run this previously on a 1TB dataset and not run these after I rebased on main. I've run it again on a 1GB dataset but had to explicitly set some Bloom filter configuration options to simulate the conditions of the larger data set. Otherwise the settings are the same as the defaults we expect most users to have.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @parthchandra

**Comment by parthchandra** [comment]
Looking into the ci failures ...

**Comment by parthchandra** [comment]
@andygrove could you please trigger a re-run of the failed ci workflow? (Only committers can do that). I cannot reproduce the failure locally and am pretty sure it has nothing to do with this PR.

**Comment by viirya** [comment]
Re-triggered.

**Comment by parthchandra** [comment]
Could we merge this please? @viirya @andygrove

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:178]:
```suggestion
              s"Partition schema $partitionSchema is not supported")
```

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:38]: Actually this only works on SparkPlan, I'm wondering why we use `TreeNode[_]` in many places?

File: spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:61]: breadth first and leaf nodes first seem conflicting?

File: spark/src/main/scala/org/apache/comet/shims/ShimCometSparkSessionExtensions.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/shims/ShimCometSparkSessionExtensions.scala:55]: Can we add clear reference here that when it is added to Spark, e.g., Spark xx+?

File: spark/src/main/scala/org/apache/spark/sql/ExtendedExplainGenerator.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/ExtendedExplainGenerator.scala:27]:
```suggestion
/**
 * A trait for a session extension to implement that provides addition explain plan information.
 * We copy this from Spark 4.0 since this trait is not available in Spark 3.x. We can remove this
 * after dropping Spark 3.x support.
 */
```

File: spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala
Comment (viirya) [inline spark/src/test/scala/org/apache/spark/sql/CometTPCQueryListBase.scala:92]: Why we need these `bloomFilter` configs?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1029]: Can we override this method for this kind of usage so we can do `withInfo(expr, child)`?

Can be in a follow up PR.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
There is conflict needed to be resolved. And I have a few minor comments like code comment changes.

**Comment by viirya** [comment]
Thank you @parthchandra. I will merge this once CI passes.

**Comment by parthchandra** [comment]
Thanks for the review @viirya @andygrove @advancedxy

**Comment by viirya** [comment]
Merged. Thanks all.

**Pull request merged**
